doc/catalyst-cli/catalyst-cli.rst

```diff
@@ -0,0 +1,241 @@
+""""""""""""""""""""""""""""""""""""""
+
+Keep intermediate files after each pipeline in the compilation. By default, no intermediate files
+are saved. Using ``--keep-intermediate`` is equivalent to using ``--save-ir-after-each=pipeline``.
+
+``--catalyst-pipeline=<pipeline1(pass1[;pass2[;...]])[,pipeline2(...)]>``
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Specify the Catalyst compilation pass pipelines.
+
+A pipeline is composed of a semicolon-delimited sequence of one or more transformation or
+optimization passes. Multiple pass pipelines can be specified and input as a comma-delimited
+sequence of pipelines.
+
+For example, if we wanted to specify two pass pipelines, ``pipe1`` and ``pipe2``, where ``pipe1``
+applies the passes ``split-multiple-tapes`` and ``apply-transform-sequence``, and where ``pipe2``
+applies the pass ``inline-nested-module``, we would specify this pipeline configuration as:
+
+.. code-block::
+
+    --catalyst-pipeline=pipe1(split-multiple-tapes;apply-transform-sequence),pipe2(inline-nested-module)
+
+``--workspace=<path>``
+""""""""""""""""""""""
+
+The workspace directory where intermediate files are saved. The default is the current working
+directory.
+
+Note that the workspace directory must exist before running ``catalyst-cli`` with this option.
+
+``--module-name=<name>``
+""""""""""""""""""""""""
+
+The module name used in naming the output file(s). The default is ``"catalyst_module"``. Using the
+``-o`` option to specify the output filename overrides this option.
+
+``--async-qnodes[=<true|false>]``
+"""""""""""""""""""""""""""""""""
+
+Enable asynchronous QNodes.
+
+``--checkpoint-stage=<stage name>``
+"""""""""""""""""""""""""""""""""""
+
+Define a *checkpoint stage*, used to indicate that the compiler should start only after reaching the
+given pass.
+
+``--dump-catalyst-pipeline[=<true|false>]``
+"""""""""""""""""""""""""""""""""""""""""""
+
+Print (to stderr) the pipeline(s) that will be run.
+
+Examples
+^^^^^^^^
+
+To illustrate how to use the Catalyst CLI tool, consider the simple MLIR code, ``my_circuit.mlir``,
+which defines a function ``my_circuit`` that implements a single-qubit quantum circuit that applies
+the sequence of gates :math:`R_x(\theta) \to H \to H \to R_x(\theta)` to the input qubit for some
+rotation angle :math:`\theta`:
+
+.. code-block:: mlir
+
+    module {
+      func.func @my_circuit(%in_qubit: !quantum.bit, %angle: f64) -> !quantum.bit {
+        %0 = quantum.custom "RX"(%angle) %in_qubit : !quantum.bit
+        %1 = quantum.custom "Hadamard"() %0 : !quantum.bit
+        %2 = quantum.custom "Hadamard"() %1 : !quantum.bit
+        %3 = quantum.custom "RX"(%angle) %2 : !quantum.bit
+        return %3 : !quantum.bit
+      }
+    }
+
+We'll use the Catalyst CLI tool to run the ``quantum-opt`` compiler to perform the MLIR-level
+optimizations and lower the input to the LLVM MLIR dialect. We'll define a pass pipeline that
```

Comment:
Doesn't the `opt` only run the passes and not lower to dialects/object files? 

(Also it's mlir llvm dialect, not llvm mlir dialect; mlir is the one that has the dialects and one of the dialects is called the "llvm dialect")

Reply:
Maybe "lower" is the wrong word, rather it "translates" from one MLIR dialect to another—in this case the output is in the LLVM dialect.

Now we're getting into the nitty-gritty language details :wink:. To me, "MLIR dialect" is the noun, and here "LLVM" acts as an adjective that modifies this noun. Want kind of MLIR dialect is it? It's the _LLVM_ MLIR dialect. Writing "MLIR LLVM dialect" reads like there's a thing called an "LLVM dialect" that is of the "MLIR" variety, which sounds strange to my ears.

The official docs call it the "LLVM dialect" or the "LLVM IR dialect" to distinguish it from true LLVM IR: https://mlir.llvm.org/docs/Dialects/LLVM/. Maybe we should follow this style as well?

Reply:
Discusses offline, we will call it the "LLVM Dialect" (as per mlir official docs) and link to the mlir docs of the dialect

Reply:
Done in c7f7528e